src/slam_toolbox_common.cpp

```diff
@@ -600,6 +606,34 @@ LocalizedRangeScan * SlamToolbox::addScan(
   return range_scan;
 }
 
+/*****************************************************************************/
+bool SlamToolbox::setLocalizationModeCallback(
+  const std::shared_ptr<rmw_request_id_t> request_header,
+  const std::shared_ptr<std_srvs::srv::SetBool::Request> req,
+  std::shared_ptr<std_srvs::srv::SetBool::Response> resp)
+/*****************************************************************************/
+{
+  bool in_localization_mode = processor_type_ == PROCESS_LOCALIZATION;
+  if (req->data != in_localization_mode) {
+    if (!req->data) {
+      // clear localization buffer
+      boost::mutex::scoped_lock lock(smapper_mutex_);
+      RCLCPP_INFO(get_logger(), "Clearing localization buffer.");
+      smapper_->clearLocalizationBuffer();
+
+      RCLCPP_INFO(get_logger(), "Entering mapping mode.");
+      processor_type_ = PROCESS;
+    }
+    else {
+      RCLCPP_INFO(get_logger(), "Entering localization mode.");
+      processor_type_ = PROCESS_LOCALIZATION;
```

Comment:
Doesn't this need to set an initial pose?

Reply:
It doesn't seem like anything like that is needed.  As far as I can tell, the only difference between `Mapper::Process` and `Mapper::ProcessLocalization` is the addition of this line:
```
  AddScanToLocalizationBuffer(pScan, scan_vertex);
```

So toggling between mapping and localization should just control if a scan gets added to the localization buffer (and potentially the oldest node in the buffer getting popped off of the buffer and removed graph) or if the localization buffer is cleared and not used.

Reply:
Toggling between the two modes works as expected in my testing.

Reply:
The localization/mapping nodes has different services setup, so which would this be able to process with? Both? If you called this in mapping mode, then there would be no `initialpose` subscriber to allow a user to specify initial poses (or correct poses) or clear localization buffer https://github.com/SteveMacenski/slam_toolbox/blob/518378e716fda74b34395265ee17c3ad7bb9e0c1/src/slam_toolbox_localization.cpp#L33-L40. If run in localization mode, mapping would have no clear queue service https://github.com/SteveMacenski/slam_toolbox/blob/518378e716fda74b34395265ee17c3ad7bb9e0c1/src/slam_toolbox_sync.cpp#L31

Reply:
Ah, I see.  The use case I had in mind was for running the non-localization nodes, but then being able to dynamically toggle on localization live.  So the position would already be known at toggle time and no initial position is required.

I hadn't thought about this service also being advertised for the localization node since it's being added to the base class.  It probably shouldn't be advertised in the localization node case.  The localization node `addScan` method also isn't set up to handle `processor_type_ == PROCESS`

Maybe I can move the service setup into the  `AsynchronousSlamToolbox`, `SynchronousSlamToolbox`, and `LifelongSlamToolbox` constructors?

Reply:
The clear localization buffer is probably useful.   I could move that to the base class.

Reply:
I disagree, these are intentionally separated. Lots of users had problems that they thought they could use initial pose in mapping mode and were filing tickets because it "didn't work" without understanding.

Reply:
Creating a new node is fine with me if you think it will be confusing.

I'm not sure what your disagreeing with, though.   I'm not proposing to add initial pose to any of the mapping nodes, just to allow the localization buffer to be toggled on and off for those nodes, plus adding the service to clear the localization buffer.

Reply:
But another important interface is the relocalization topic `initialpose` that would also be highly relevent to users if toggling between mapping and localization. That's what I really don't want to have show up for SLAM users and would be necessary in localization mode

Reply:
Ok.  Can you explain how the `initialpose` topic works?  It wasn't clear to me.  

Is it used in a case where the position estimate diverges from the map, maybe by exploring an area not yet mapped. Then, when it returns to the already mapped area, it is to far diverged for a automatic loop closure to be detected?

Reply:
Its analog to the use in AMCL: you can publish a pose to that topic and it relocalizes the robot to that position. So its only usable in localization mode but can be used to set the initial pose on startup or correct a pose from delocalization

Reply:
I see.  That makes sense.  In the startup case, is it also possible to do this by:
1.  providing a pose in the `deserialize_map` service?
2.  or providing `map_file_name` and `map_start_pose` launch params?

Reply:
It sounds like it's best to create a new node.  `slam_toolbox_mixed` works for me, though its not the most intuitive.  Not sure what would be better.  If it's a new node, would it be ok to break the AMCL `initialpose` topic convention and provide a service instead?  Like `set_localization_pose`?

Reply:
[Yes](https://github.com/SteveMacenski/slam_toolbox/blob/ros2/srvs/DeserializePoseGraph.srv#L11) and [Yes](https://github.com/SteveMacenski/slam_toolbox/blob/71c1c457d0be4f86b07e3026fc986fa41afb630b/src/slam_toolbox_common.cpp#L317) but its not only useful in startup. Its also useful for delocalization or relocalization after a person physically moves the robot. 

It should be that topic so that its analog with all other localization methods. Its not a written rule anywhere, but `initialpose` topic is used by virtually all localization engines in ROS to relocalize a robot. This is probably established from AMCL, but the rviz tools and other localization algorithms in the ROS ecosystem all do this, which is convenient for cross compatibility

Reply:
@SteveMacenski Ok.  I reworked this to be a separate node in 'experimental'.   The new mode is a subclass of the Localization mode, but redefines the relevant functions to either use the Localization implementation or the Common (mapping) implementation based on if it is in "localization" or "mapping" mode.

I called in 'map_and_localization', though I'm open to naming it something else.